### PR TITLE
feat: add support for polls

### PR DESCRIPTION
**Pull request by Mateo-tem**
**Please describe the changes this PR makes and why it should be merged:**

**If applicable, please reference Discord API Docs PRs or commits that influenced this PR:**
- https://github.com/discord/discord-api-docs/pull/6746

### CODE REVIEW

File: deno/gateway/v10.ts [hunk 1]
```diff
@@ -2051,6 +2053,33 @@ export interface GatewayPresenceUpdateData {
...
+	guild_id?: Snowflake;
+	/**
+	 * ID of the answer
+	 */
+	answer_id: number;
+}
+
+export type GatewayMessagePollVoteAdd = GatewayMessagePollVoteData;
+
+export type GatewayMessagePollVoteRemove = GatewayMessagePollVoteData;
```
Comment (vladfrangu) [inline deno/gateway/v10.ts:2081]: These need to be added to GatewayDispatchPayload and follow the structure of all other payloads

File: deno/payloads/v10/poll.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	/**
+	 * The results of the poll
+	 */
+	results: APIPollResults;
+}
+
+/**
+ * https://discord.com/developers/docs/resources/poll#layout-type
+ */
+export enum APIPollLayoutType {
```
Comment (vladfrangu) [inline deno/payloads/v10/poll.ts:40]: Enums do not have API prefix

File: deno/payloads/v10/poll.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+ */
+export interface APIPollResults {
+	/**
+	 * Whether the votes have been precisely counted
+	 */
+	is_finalized: boolean;
+	/**
+	 * The counts of each answer
+	 */
+	answer_counts: APIPollAnswerCount;
```
Comment (vladfrangu) [inline deno/payloads/v10/poll.ts:86]: Isn't this an array?

**Review by vladfrangu — CHANGES_REQUESTED** [review]
Also need to add the routes in the rest files

File: rest/v10/index.ts [hunk 4]
```diff
@@ -461,6 +462,22 @@ export const Routes = {
 		return `/guilds/${guildId}/templates/${code}` as const;
 	},
 
+	/**
+	 * Route for:
+	 * - GET `/channels/{channel.id}/poll/{message.id}/answers/{answer_id}`
+	 */
+	pollAnswerVoters(channelId: Snowflake, messageId: Snowflake, answerId: string) {
+		return `/channels/${channelId}/poll/${messageId}/answers/${answerId}` as const;
```
Comment (almeidx) [inline rest/v10/index.ts:470]: The routes were correct before 2c4b8acb2ae81a4afd4678c2597cb1f06e469614. It's `polls`, not `poll`

**Comment by vladfrangu** [comment]
This needs a rebase

File: deno/gateway/v10.ts [hunk 5]
```diff
@@ -2051,6 +2055,45 @@ export interface GatewayPresenceUpdateData {
  */
 export type GatewayActivityUpdateData = Pick<GatewayActivity, 'name' | 'state' | 'type' | 'url'>;
 
+export interface GatewayMessagePollVoteData {
```
Comment (vladfrangu) [inline deno/gateway/v10.ts:2058]: This is in the wrong region, this should be moved in the received payloads
Also needs to follow the name format (and ideally export aliases for the event name and dispatch data

File: deno/payloads/v10/poll.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+/**
+ * https://discord.com/developers/docs/resources/poll#poll-answer-object-poll-answer-object-structure
+ */
+export interface APIPollAnswer {
+	/**
+	 * The ID of the answer. Starts at `1` for the first answer and goes up sequentially
+	 *
+	 * Only sent as part of responses from Discord's API/Gateway
```
Comment (vladfrangu) [inline deno/payloads/v10/poll.ts:70]: Does this mean it can be undefined?

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: deno/payloads/v10/poll.ts (hunk 6 above)
Comment (Mateo-tem) [reply deno/payloads/v10/poll.ts:70]: Uhmm, no. It means users cannot send an answer ID within an answer object to create a poll. I think is a bit obvious, not sure if I should leave it that way

File: payloads/v10/poll.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+	 */
+	allow_multiselect: boolean;
+	/**
+	 * The layout type of the poll
+	 */
+	layout_type: PollLayoutType;
+	/**
+	 * The results of the poll
+	 */
+	results: APIPollResults;
```
Comment (almeidx) [inline payloads/v10/poll.ts:34]: This is optional

File: payloads/v10/poll.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+/**
+ * https://discord.com/developers/docs/resources/poll#poll-answer-object-poll-answer-object-structure
+ */
+export interface APIPollAnswer {
+	/**
+	 * The ID of the answer. Starts at `1` for the first answer and goes up sequentially
+	 *
+	 * Only sent as part of responses from Discord's API/Gateway
```
Comment (almeidx) [inline payloads/v10/poll.ts:70]: Let's get rid of this and change the `RESTAPIPollCreate` interface to not have this prop

File: rest/common.ts [hunk 9]
```diff
@@ -295,6 +295,14 @@ export enum RESTJSONErrorCodes {
 	CannotUpdateOnboardingWhileBelowRequirements,
 
 	FailedToBanUsers = 500_000,
+
+	PollVotingBlocked = 520_000,
+	PollExpired,
+	InvalidChannelTypeForPollCreation,
+	CannotEditAPollMessage,
+	CannotUseAnEmojiIncludedWithAPoll,
```
Comment (almeidx) [inline rest/common.ts:303]:
```suggestion
	CannotUseAnEmojiIncludedWithThePoll,
```

File: rest/v10/index.ts [hunk 10]
```diff
@@ -461,6 +462,22 @@ export const Routes = {
 		return `/guilds/${guildId}/templates/${code}` as const;
 	},
 
+	/**
+	 * Route for:
+	 * - GET `/channels/{channel.id}/polls/{message.id}/answers/{answer_id}`
+	 */
+	pollAnswerVoters(channelId: Snowflake, messageId: Snowflake, answerId: string) {
```
Comment (almeidx) [inline rest/v10/index.ts:469]:
```suggestion
	pollAnswerVoters(channelId: Snowflake, messageId: Snowflake, answerId: number) {
```

**Review by almeidx — CHANGES_REQUESTED** [review]

File: rest/v10/poll.ts [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+	 *
+	 * @default 25
+	 */
+	limit?: number;
+}
+
+/**
+ * https://discord.com/developers/docs/resources/poll#poll-create-request-object-poll-create-request-object-structure
+ */
+export interface RESTAPIPollCreate extends Omit<APIPoll, 'answer_id' | 'expiry' | 'results'> {
```
Comment (almeidx) [inline rest/v10/poll.ts:23]: `answer_id` is not right here. You have to change the `answers` property instead

**Review by almeidx — CHANGES_REQUESTED** [review]

**Review by almeidx — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Pull request merged**